### PR TITLE
Keypresses

### PR DESCRIPTION
**Pull request by zcorpan**
API to simulate keypresses. Milestone 3 in #15

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria-at-automation/pull/26.html" title="Last updated on Nov 22, 2022, 12:11 AM UTC (8009bfc)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria-at-automation/26/9faba32...8009bfc.html" title="Last updated on Nov 22, 2022, 12:11 AM UTC (8009bfc)">Diff</a>

### CODE REVIEW

File: index.bs
Comment (jkva) [inline index.bs:286]: Given each `Error code` is identical to its corresponding `JSON error code`, is this distinction necessary? Can we consider hypotheticals where this is not the case?

**Review by jkva — CHANGES_REQUESTED** [review]
Consider this review to also encompass the previous reviews requested for PRs, since these all incorporate each other.

The `extract-schemas.js` script is currently invalid because of added CDDL; see comments.

**Comment by jkva** [comment]
In order to understand the extraction script better, I've separated some code out into functions. Feel free to incorporate what seems useful to you.


```js
import { Lexer, Parser } from 'cddl';
import * as fs from 'node:fs/promises';

const parseCddl = input => new Parser(new Lexer(input)).parse();

const cddlToJSONSchema = input => {
  const cddlAst = parseCddl(input);
  const jsonSchemaAst = {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "TODO",
    "$defs": {},
  };
  for (const obj of cddlAst) {
    console.log(JSON.stringify(obj, null, 2)); // TODO
    // For SessionNewCommand, need to not set additionalProperties: false
    // since Command allows additional properties, all schemas in Command's allOf
    // need to also allow additional properties (otherwise 'id' will be invalid).
  }
}

const removeIndentation = (indentation, cddl) => {
  if (indentation.length === 0) {
    return cddl;
  }

  let indentationRegexp = new RegExp(`^${indentation}`);
  return cddl
      .split('\n')
      .map(line => line.replace(indentationRegexp, ''))
      .join('\n');
}

const formatCddl = cddl => cddl.join('\n\n').trim() + '\n';

const extractCddlFromSpec = async () => {
  const source = await fs.readFile('index.bs', { encoding: 'utf8' });
  const matches = [...source.matchAll(/^([ \t]*)<pre class=['"]cddl((?: [a-zA-Z0-9_-]+)+)['"]>([\s\S]*?)<\/pre>/gm)];

  const [local, remote] = matches.reduce(([local, remote], match) => {
    const [_, indentation, cssClass, content] = match;

    let isLocal = cssClass.indexOf(' local-cddl') > -1;
    let isRemote = cssClass.indexOf(' remote-cddl') > -1;

    if (!isLocal && !isRemote) {
      return [local, remote];
    }

    let cddl = removeIndentation(indentation, content.trim());

    if (isLocal) {
      local.push(cddl);
    } else {
      remote.push(cddl);
    }

    return [local, remote];
  }, [[], []])

  return [
      formatCddl(local),
      formatCddl(remote)
  ]
}

try {
  await fs.mkdir('schemas');
} catch(ex) {
  if (ex.code !== 'EEXIST') {
    throw ex;
  }
}

const [localCddl, remoteCddl] = await extractCddlFromSpec();

await fs.writeFile('schemas/at-driver-local.cddl', localCddl);
await fs.writeFile('schemas/at-driver-remote.cddl', remoteCddl);

// Convert CDDL to JSON Schema
cddlToJSONSchema(localCddl);
cddlToJSONSchema(remoteCddl);
```

**Comment by jkva** [comment]
@zcorpan I noticed you also ran into a problem with the module entry script being incorrect; I've submitted [a PR which addresses this](https://github.com/saucelabs/cddl/pulls/5). Yet if this is not incorporated by the maintainer, we should consider forking the module.

File: index.bs
Comment (jugglinmike) [reply index.bs]: Thanks for pointing this out! Simon's going to be unavailable for a bit, but I can lend a hand in his absence. Your input prompted me to check the spelling of this document more generally, and I found a handful of other typos. I've corrected those in gh-19, where they were originally submitted. So double thanks!

File: scripts/extract-schemas.js
Comment (jugglinmike) [reply scripts/extract-schemas.js:4]: That seems like a fair assumption to me. I've removed it in order to limit further confusion--see gh-19.

File: index.bs
Comment (jugglinmike) [reply index.bs:640]: By my reading of [rfc8610](https://datatracker.ietf.org/doc/html/rfc8610), the proposed syntax is valid CDDL, and if that's so, then the Node.js package's behavior is a deficiency.

Addressing the deficiency by limiting this proposal's usage of CDDL sounds like the lowest-friction solution. At the same time, doing so would hide a rough edge for future work since contributors will not necessarily recognize that the full language is not available to them, and they may not be willing/able to deviate from their structure of choice in order to satisfy a tool. It may be better to extend the utility or find an alternative that's more complete. That depends on factors which are beyond my perspective at the moment, so I'm going to hold off on this for the time being.

**Comment by jugglinmike** [comment]
> @zcorpan I noticed you also ran into a problem with the module entry script being incorrect; I've submitted [a PR which addresses this](https://github.com/saucelabs/cddl/pulls/5). Yet if this is not incorporated by the maintainer, we should consider forking the module.

Much appreciated. I'm including a corrected version of the link so that GitHub recognizes the relationship between this pull request and that one:

https://github.com/saucelabs/cddl/pull/5

And in the mean time, the following change will allow the script in this patch to run (though I don't recommend relying on it indefinitely):

```diff
diff --git a/scripts/extract-schemas.js b/scripts/extract-schemas.js
index 0bd627d..582e0fc 100644
--- a/scripts/extract-schemas.js
+++ b/scripts/extract-schemas.js
@@ -1,4 +1,4 @@
-import { Lexer, Parser } from 'cddl';
+import { Lexer, Parser } from 'cddl/build/index.js';
 import * as fs from 'node:fs/promises';
```

File: index.bs
Comment (zcorpan) [reply index.bs:286]: I did it this way to match [WebDriver](https://w3c.github.io/webdriver/#dfn-error-code), except we don't have HTTP response codes.

The table makes the casing explicit, at least, so I think we should keep it even if the names and the strings always match.

File: index.bs
Comment (zcorpan) [reply index.bs:640]: I've worked around the bug for now and filed an issue: https://github.com/saucelabs/cddl/issues/7

**Comment by zcorpan** [comment]
Thank you for your review @jkva ! I've incorporated your script refactoring in #19 and I think all comments are now addressed.

I would like to merge each open PR in order so that the history is clear both in the commit log and in GitHub's UI. We've set up this repo to require an approved review in GitHub before merging PRs. Since you have reviewed this, it means #19 and #25 are also reviewed. Can you mark this PR and the other 2 as reviewed, please?

The Settings PR is not included here: #22. If you have reviewed that one, can you also approve it?

Thanks!

File: index.bs
Comment (zcorpan) [reply index.bs]: Thanks, fixed.

**Comment by s3ththompson** [comment]
@jkva could you take another look and mark as reviewed if this looks good to merge? thanks!

File: index.bs
Comment (jkva) [inline index.bs:654]: How does this `ActionItem` account for meta-keys?

**Review by jkva — COMMENTED** [review]
Not per se related to this PR, but I notice that lines 221 and 222 of `index.bs` refer to the key `platform` while it seems that ought to be `platformName`. 

There are several lines where there is referral to a definition having been defined in CDDL. This seems redundant and perhaps ought to be written as a definition existing in as CDDL.

**Comment by zcorpan** [comment]
> Not per se related to this PR, but I notice that lines 221 and 222 of `index.bs` refer to the key `platform` while it seems that ought to be `platformName`.

Good catch! Fix: https://github.com/w3c/aria-at-automation/pull/29

> There are several lines where there is referral to a definition having been defined in CDDL. This seems redundant and perhaps ought to be written as a definition existing in as CDDL.

Hmm, yeah. Filed a new issue: https://github.com/w3c/aria-at-automation/issues/30

**Comment by zcorpan** [comment]
I've now updated the PR to specify `interaction.pressKeys` as discussed in https://github.com/w3c/aria-at-automation/pull/26#issuecomment-1178106223

>  Include a special modifier for the screen-reader specific modifier keys.

This is not done yet. How do we want to support this? Should there be a special string for each vendor in place of the raw key string, e.g. `"nvda"`, `"macOS VoiceOver"`, etc?

```
{
  "method": "interaction.pressKeys",
  "params": {
    "keys": ["nvda", "a"]
  }
}
```

Or a boolean property in `InteractionPressKeysParameters` to indicate that the screen reader specific modifier keys should be pressed?

```
{
  "method": "interaction.pressKeys",
  "params": {
    "keys": ["a"],
    "vendorModifier": true
  }
}
```

**Comment by zcorpan** [comment]
I filed a new issue #34 to track the above question.

As discussed in our call today, we could merge this PR without special modifier support and address that later.

**Review by jkva — APPROVED** [review]

**Review by jugglinmike — APPROVED** [review]
The final remote end step describes a pretty involved algorithm. I think it'd be more clear as a series of steps, particularly if the "implementation defined" part was formally structured as its own algorithm (e.g. "simulate keyboard interaction"). Given that this has already been accepted in its current form, though, I'll propose the refactoring in a follow-up patch.

**Comment by jugglinmike** [comment]
Here's that patch: https://github.com/w3c/aria-at-automation/pull/41

**Pull request merged**
